File: cip/src/formats/epub_reader.cpp
// Copyright (c) 2022 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

#include "formats/epub_reader.h"

#include <QDebug>

EpubReader::EpubReader(QObject* parent) : QObject(parent) {

}

EpubReader::~EpubReader() {
  // TODO(Shaohua):
  // epub_cleanup();

  this->cleanup();
}

void EpubReader::cleanup() {
  if (it_ != nullptr) {
    epub_free_iterator(it_);
    it_ = nullptr;
  }

  if (epub_ != nullptr) {
    epub_close(epub_);
    epub_ = nullptr;
  }
}

bool EpubReader::load(const QString& filepath) {
  this->cleanup();

  epub_ = epub_open(filepath.toStdString().c_str(), 0);
  if (epub_ == nullptr) {
    qWarning() << "Failed to open epub file:" << filepath;
    return false;
  }

//  epub_dump(epub_);

  return true;
}

bool EpubReader::initIterator() {
  if (epub_ == nullptr) {
    qWarning() << "epub file not loaded";
    return false;
  }

  if (it_ != nullptr) {
    epub_free_iterator(it_);
  }
  it_ = epub_get_iterator(epub_, EITERATOR_SPINE, 0);
  if (it_ == nullptr) {
    qWarning() << "Failed to get eit of epub file";
    return false;
  }

  return true;
}

int EpubReader::numPages() {
  if (!this->initIterator()) {
    return -1;
  }

  char* content = epub_it_get_curr(it_);
  int pages = -1;
  while (content != nullptr) {
    content = epub_it_get_next(it_);
    pages += 1;
  }

  return pages;
}

bool EpubReader::readPage(int number, QString& text) {
  if (!this->initIterator()) {
    return false;
  }

  char* content = nullptr;
  int current = -1;
  while (current < number){
    content = epub_it_get_next(it_);
    if (content == nullptr) {
      break;
    }
    current += 1;
  }
  if (content == nullptr) {
    return false;
  }
  text = content;
  return true;
}

File: cip/src/formats/epub_reader.rs
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

use epub::doc::EpubDoc;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;

use crate::error::{Error, ErrorKind};

pub struct EpubReader {
    doc: EpubDoc<BufReader<File>>,
}

impl EpubReader {
    /// Open an epub file.
    ///
    /// # Errors
    /// Returns error if epub file does not found or is invalid.
    pub fn open<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
        let doc = EpubDoc::new(path)?;
        Ok(Self { doc })
    }

    #[must_use]
    pub fn pages(&self) -> usize {
        self.doc.get_num_pages()
    }

    /// Read text content of specific page.
    ///
    /// # Errors
    /// Returns error if failed to read page text.
    pub fn read_page(&self, page: usize) -> Result<String, Error> {
        if page < self.doc.spine.len() {
            Ok(self.doc.spine[page].clone())
        } else {
            Err(Error::from_string(
                ErrorKind::InvalidEpubPage,
                "Invalid page number".to_string(),
            ))
        }
    }
}
